network/network-address/src/encrypted.rs

```diff
@@ -0,0 +1,482 @@
+    type Error = lcs::Error;
+
+    fn try_from(value: &EncNetworkAddress) -> Result<Self, lcs::Error> {
+        let bytes = lcs::to_bytes(value)?;
+        Ok(RawEncNetworkAddress::new(bytes))
+    }
+}
+
+#[cfg(any(test, feature = "fuzzing"))]
+impl Arbitrary for RawEncNetworkAddress {
+    type Parameters = ();
+    type Strategy = BoxedStrategy<Self>;
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        any::<EncNetworkAddress>()
+            .prop_map(|enc_addr| RawEncNetworkAddress::try_from(&enc_addr).unwrap())
+            .boxed()
+    }
+}
+
+///////////////////////
+// EncNetworkAddress //
+///////////////////////
+
+impl EncNetworkAddress {
+    pub fn encrypt(
+        addr: RawNetworkAddress,
+        root_key: &RootKey,
+        root_key_version: RootKeyVersion,
+        account: &AccountAddress,
+        seq_num: u64,
+        addr_idx: u32,
+    ) -> Result<Self, Error> {
+        // unpack the RawNetworkAddress into its base Vec<u8>
+        let mut addr_vec: Vec<u8> = addr.into();
+
+        let account_key = Self::derive_account_key(root_key, account)?;
+        let aead = Aes256Gcm::new(GenericArray::from_slice(&account_key));
+
+        // nonce := seq_num || addr_idx; sizeof(nonce) == 12
+        // ex: seq_num = 0x1234, addr_idx = 0x04
+        //     ==> nonce_slice == &[0, 0, 0, 0, 0, 0, 0x12, 0x34, 0, 0, 0, 0x4]
+        let nonce = ((seq_num as u128) << 32) | (addr_idx as u128);
+        let nonce_buf = (nonce as u128).to_be_bytes();
+        let nonce_slice = &nonce_buf[mem::size_of::<u128>() - AES_GCM_NONCE_LEN..];
+        let nonce_slice = GenericArray::from_slice(nonce_slice);
+
+        // the root_key_version is in-the-clear, so we include it in the integrity check
+        // using the "additonal data"
+        let ad_buf = root_key_version.to_be_bytes();
+        let ad_slice = &ad_buf[..];
+
+        // encrypt the raw network address in-place
+        let auth_tag = aead
+            .encrypt_in_place_detached(nonce_slice, ad_slice, &mut addr_vec)
+            .map_err(|_| Error::EncryptError)?;
+
+        // append the authentication tag
+        addr_vec.extend_from_slice(auth_tag.as_slice());
+
+        Ok(Self {
+            root_key_version,
+            seq_num,
+            enc_addr: addr_vec,
+        })
+    }
+
+    pub fn decrypt(
+        self,
+        root_key: &RootKey,
+        account: &AccountAddress,
+        addr_idx: u32,
+    ) -> Result<RawNetworkAddress, Error> {
+        let root_key_version = self.root_key_version;
+        let seq_num = self.seq_num;
+        let mut enc_addr = self.enc_addr;
+
+        // ciphertext is too small to even contain the authentication tag, so it
+        // must be invalid.
+        if enc_addr.len() < AES_GCM_TAG_LEN {
+            return Err(Error::CiphertextTooSmall(enc_addr.len()));
+        }
+
+        let account_key = Self::derive_account_key(root_key, account)?;
```

Comment:
IIUC this will never fail, but I can agree that it's better to just bubble up the error if we're already returning a Result anyway